### PR TITLE
Review: feat: be able to change the destination of each created file

### PR DESCRIPTION
**Pull request by tdurieux**
What do you think?

The idea of this is to be able to override DefaultOutputDestination and defines your output strategy in `getOutputPath(File defaultOutputDirectory, Path packagePath, String filename)`

### CODE REVIEW

File: src/main/java/spoon/support/OutputDestination.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import spoon.reflect.declaration.CtPackage;
+import spoon.reflect.declaration.CtType;
+
+import java.io.File;
+import java.nio.file.Path;
+
+public interface OutputDestination {
+
+	Path getOutputPath(File defaultOutputDirectory, Path packagePath, String filename);
```
Comment (monperrus) [inline src/main/java/spoon/support/OutputDestination.java]: Path, Path?

File: src/main/java/spoon/support/OutputDestination.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+	Path getOutputPath(CtPackage type, File defaultOutputDirectory,
```
Comment (monperrus) [inline src/main/java/spoon/support/OutputDestination.java]: additional arg at the end?

**Comment by pvojtechovsky** [comment]
I like that idea.
Note: Interface OutputDestination with 3 methods cannot be implemented using lambda. May be you can implement two of them as `default` methods directly in interface?

**Comment by tdurieux** [comment]
I think three methods is too much.
I propose to only have
`getOutputPath(CtElement element, File defaultOutputDirectory, Path packagePath, String filename)`.

The users that will use this interface should be able to do an instanceof .

WDYT?

**Comment by pvojtechovsky** [comment]
What about method like this:
```
Path getOutputPath(CtPackage package, CtType type, File defaultOutputDirectory, Path packagePath, String filename)
```
the `type` might be sometime null

**Comment by monperrus** [comment]
ping?

**Comment by tdurieux** [comment]
I have an issue because the output destination is stored in two places in the:

* ModelBuilder
* FileGenerator

I think it is an issue, but I don't know which one is the best to keep

**Comment by surli** [comment]
> I think it is an issue, but I don't know which one is the best to keep

I propose to put the output destination in Environment and to only store it there. Then other classes could get it through Environment. WDYT?

**Comment by pvojtechovsky** [comment]
>to put the output destination in Environment

sounds good for me

**Comment by monperrus** [comment]
rebase now that we have #1770 ?

**Comment by monperrus** [comment]
travsi fails? @tdurieux ?

**Comment by tdurieux** [comment]
no there is still a lot of work to do.
The current implementation is not nice, way to much state

**Comment by tdurieux** [comment]
I finished to clean up this PR.

But I'm not sure if the interface and the implementation of `OutputDestination` is good enough.
I made `DefaultOutputDestination` in the way that it is easy to overwrite.
Probably, the method that will be overwritten by the client is `Path getDirectoryPath(CtModule module, CtPackage pack, CtType type)` which returns the root path of the destination.

File: src/main/java/spoon/compiler/Environment.java [hunk 3]
```diff
@@ -344,6 +345,16 @@ void report(Processor<?> processor, Level level,
 	 */
 	File getSourceOutputDirectory();
 
+	/**
+	 * Set the output destination that handles where source files are written
+	 */
+	void setOutputDesination(OutputDestination outputDestination);
```
Comment (surli) [inline src/main/java/spoon/compiler/Environment.java]: The name is a bit confusing with `sourceOutputDirectory`, and for me it looks like a strategy for the final location of the output. So WDYT about renaming it to something like `setOutputStrategy` or `setOuputHandler`?

File: src/main/java/spoon/support/JavaOutputProcessor.java [hunk 4]
```diff
@@ -217,38 +217,32 @@ private void createModuleFile(CtModule module) {
...
-		if (pack.isUnnamedPackage()) {
-			packageDir = this.getModuleFile(pack.getDeclaringModule());
-		} else {
-			// Create current package dir
-			packageDir = new File(this.getModuleFile(pack.getDeclaringModule()).getAbsolutePath() + File.separatorChar + pack.getQualifiedName().replace('.', File.separatorChar));
-		}
-		if (!packageDir.exists()) {
-			if (!packageDir.mkdirs()) {
+	private Path getElementPath(CtPackage type) {
+		return createFolders(getEnvironment().getOutputDestination()
```
Comment (surli) [inline src/main/java/spoon/support/JavaOutputProcessor.java]: Why don't you delegate the creation of the folder to the output destination strategy? We can imagine different implementation e.g. a strategy for not creating new directory, or for not overwritten existing files.

File: src/main/java/spoon/compiler/Environment.java (hunk 3 above)
Comment (tdurieux) [reply src/main/java/spoon/compiler/Environment.java]: I agree

File: src/main/java/spoon/support/JavaOutputProcessor.java (hunk 4 above)
Comment (surli) [reply src/main/java/spoon/support/JavaOutputProcessor.java]: > it is not exactly the same responsibility, what you ask is more something like, isToPrint, not whereToPrint

In fact I see the `OutputDestinationStrategy` as the main handler in Spoon to manage the pathes of the output. In that sense, its responsibility is indeed, not only to say where to print, but also to say if it's ok to print things at this specific location. 
The idea behind it is that you only need to provide your own implementation of `OutputDestinationStrategy` whenever you need a specific behaviour related to path, beeing where, if or how.

Comment (tdurieux) [reply src/main/java/spoon/support/JavaOutputProcessor.java]: I don't know how to communicate with the output processor. The use case is not clear to me.

Comment (surli) [reply src/main/java/spoon/support/JavaOutputProcessor.java]: Maybe just rename `getOutputPath` to `createOutputPath`, and inside this method you create the needed folders?..

**Comment by surli** [comment]
I changed some names so for me it's ready to merge @tdurieux @monperrus

**Comment by monperrus** [comment]
That's good. Could you add a test case for this new feature?

**Comment by surli** [comment]
test added @monperrus

**Comment by monperrus** [comment]
thanks a lot!

**Pull request merged**